verboseTree-maven-plugin/src/main/java/org/apache/maven/dependency/graph/SerializeGraph.java

```diff
@@ -0,0 +1,118 @@
+/*
+ *  Copyright 2020 Google LLC.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package org.apache.maven.dependency.graph;
+
+import org.eclipse.aether.artifact.Artifact;
+import org.eclipse.aether.graph.DependencyNode;
+
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Set;
+
+public class SerializeGraph
+{
+    // will be injected eventually
+    private String outputType;
+
+    private final Map<DependencyNode, Boolean> visitedNodes;
+    private final Set<String> artifactSet;
+    private StringBuilder builder;
+
+    public SerializeGraph()
+    {
+        visitedNodes = new IdentityHashMap<DependencyNode, Boolean>( 512 );
+        artifactSet = new HashSet<String>();
+        builder = new StringBuilder();
+    }
+
+    public String serialize( DependencyNode root )
+    {
+        return dfs( root, "" ).toString();
+    }
+
+    private static void appendDependency( StringBuilder builder, DependencyNode node )
+    {
+        String scope = node.getDependency().getScope();
+        builder.append( getCoordinateString( node.getArtifact() ) );
+
+        if ( scope != null && !scope.isEmpty() )
+        {
+            builder.append( ":" ).append( scope );
+        }
+    }
+
+    private static String getCoordinateString( Artifact artifact )
+    {
+        return artifact.getGroupId() + ":" + artifact.getArtifactId() + ":" +
+                artifact.getExtension() + ":" + artifact.getVersion();
```

Comment:
Do you need classifier in here?

Reply:
Also Scope? Optional? How do you determine if this comparison logic is thorough?

Reply:
Based on the documentation https://maven.apache.org/plugins/maven-dependency-plugin/examples/resolving-conflicts-using-the-dependency-tree.html I think they just use GroupID:ArtifactID:Extension:Version:Scope so I don't think I need classifier.

Reply:
@stephaniewang526 thanks for pointing this out, the comparison logic needs some additions before it's ready.

Reply:
`classifier` should be needed. Here's an example I saw recently:
```
[WARNING] Some problems were encountered while building the effective model for com.google.cloud:google-cloud-bigquerystorage:jar:0.134.0-beta
[WARNING] 'dependencies.dependency.(groupId:artifactId:type:classifier)' must be unique: com.google.protobuf:protobuf-java:jar -> duplicate declaration of version (?) @ line 159, column 17
```

[Here](https://github.com/googleapis/java-shared-dependencies/blob/772c2a2edc19e708838190e40b8df7eefb1764dc/pom.xml#L234) is a real example where we are using `test` classifier dependency as a test dependency:
```
      <dependency>
        <groupId>com.google.cloud</groupId>
        <artifactId>google-cloud-core</artifactId>
        <version>${google.core.version}</version>
        <classifier>tests</classifier>
      </dependency>
```
Note that `google-cloud-core` appears several times in the pom of java-shared-dependencies repo (in the BOM, as a test-jar, and as a test classified dep).

You may wish to figure out what Maven considers a duplicate dependency.